src/dbnode/persist/fs/msgpack/stream_with_digest.go

```diff
@@ -0,0 +1,114 @@
+package msgpack
+
+import (
+	"errors"
+	"hash"
+	"hash/adler32"
+)
+
+var (
+	// errChecksumMismatch returned when the calculated checksum doesn't match the stored checksum
+	errChecksumMismatch = errors.New("calculated checksum doesn't match stored checksum")
+)
+
+// DecoderStreamWithDigest calculates the digest as it processes a decoder stream.
+type DecoderStreamWithDigest interface {
+	DecoderStream
+
+	// Reset resets the reader for use with a new reader.
+	Reset(stream DecoderStream)
+
+	// Digest returns the digest
+	Digest() hash.Hash32
+
+	// Validate compares the current digest against the expected digest and returns
+	// an error if they don't match.
+	Validate(expectedDigest uint32) error
+
+	// Capture provides a mechanism for manually capturing bytes to add to digest when reader is manipulated
+	// through atypical means (e.g. reading directly from the backing byte slice of a ByteReader)
+	Capture(bytes []byte) error
+
+	// Returns the decoder stream wrapped by this object
+	wrappedStream() DecoderStream
+}
+
+type decoderStreamWithDigest struct {
+	reader     DecoderStream
+	digest     hash.Hash32
+	unreadByte bool
+}
+
+func newDecoderStreamWithDigest(reader DecoderStream) DecoderStreamWithDigest {
+	return &decoderStreamWithDigest{
+		reader: reader,
+		digest: adler32.New(),
+	}
+}
+
+func (d *decoderStreamWithDigest) Read(p []byte) (n int, err error) {
+	n, err = d.reader.Read(p)
+	if n > 0 {
+		start := 0
+		if d.unreadByte {
+			d.unreadByte = false
+			start++
+		}
+		if _, err := d.digest.Write(p[start:n]); err != nil {
+			return 0, err
+		}
+	}
+	return n, err
+}
+
+func (d *decoderStreamWithDigest) ReadByte() (byte, error) {
+	b, err := d.reader.ReadByte()
+	if err == nil {
+		if d.unreadByte {
+			d.unreadByte = false
+		} else {
+			if _, err := d.digest.Write([]byte{b}); err != nil {
+				return b, err
+			}
+		}
+	}
+	return b, err
+}
+
+func (d *decoderStreamWithDigest) UnreadByte() error {
```

Comment:
Curious, what are the UnreadByte and Capture APIs used for?

Reply:
I think `UnreadByte` is used by `Peek`-like operations, but I'm not entirely sure. None of the index entry decoding code actually uses it, but I had to implement it so that `decoderStreamWithDigest` adhered to the `DecoderStream` interface.

`Capture` is used so that you can arbitrarily add bytes to the digest. There's a special case in index entry decoding where we pull the desired bytes directly from the backing byte array and then use `Seek` to advance the reader. When this is done, `Capture` should be used to update the digest. Here's where that's being done: https://github.com/m3db/m3/pull/2468/files#diff-a4220f1ade552cd435cedf33723b0f81R684